### ISSUE TITLE
Unknown error in compilation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7763]
Migrated from [rt.perl.org#33914](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=33914) (status was 'resolved')

Searchable as RT33914$

**Comment by p5pRT** [comment]
### From simonf@cshl.edu

#### Created by simonf@cshl.edu



The following modules produce this error message&#8203;:

\-\->perl \-I\. \-MA
Unknown error
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.

A\.pm
\*\*\*
package A;

use strict;

use base C;

1;

\*\*\*

C\.pm
\*\*\*
package C;

use strict;

1;

\*\*\*

As far as I can tell\, the name of the second package does not matter\.
What matters is that you omit quotes around its name in 'use base'\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl v5.8.5 in the Red Hat build system.
It is being executed now by Perl v5.8.5 - Tue Oct 12 12:44:18 EDT 2004.

Site configuration information for perl v5.8.5:

Configured by Red Hat, Inc. at Tue Oct 12 12:44:18 EDT 2004.

Summary of my perl5 (revision 5 version 8 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.4.21-14.elsmp, archname=i386-linux-thread-multi
    uname='linux tweety.build.redhat.com 2.4.21-14.elsmp #1 smp wed apr 14 18:55:47 edt 2004 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -m32 -march=i386 -mtune=pentium4 -Dversion=5.8.5 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux -Dvendorprefix=/usr -Dsiteprefix=/usr -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dinc_version_list=5.8.4 5.8.3 5.8.2 5.8.1 5.8.0'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -g -pipe -m32 -march=i386 -mtune=pentium4',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.4.2 20041006 (Red Hat 3.4.2-5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.3.3.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.8.5/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.5:
    /opt/software/perl/lib/5.8.5/i386-linux-thread-multi
    /opt/software/perl/lib/5.8.5
    /opt/software/perl/lib/i386-linux-thread-multi
    /opt/software/perl/lib/5.8.4
    /opt/software/perl/lib/5.8.3
    /opt/software/perl/lib/5.8.2
    /opt/software/perl/lib/5.8.1
    /opt/software/perl/lib/5.8.0
    /opt/software/perl/lib
    /opt/software/perl/lib/perl5/site_perl/5.8.5/i386-linux-thread-multi
    /opt/software/perl/lib/perl5/site_perl/5.8.5
    /opt/software/perl/lib/perl5/site_perl/i386-linux-thread-multi
    /opt/software/perl/lib/perl5/site_perl/5.8.4
    /opt/software/perl/lib/perl5/site_perl/5.8.3
    /opt/software/perl/lib/perl5/site_perl/5.8.2
    /opt/software/perl/lib/perl5/site_perl/5.8.1
    /opt/software/perl/lib/perl5/site_perl/5.8.0
    /opt/software/perl/lib/perl5/site_perl
    /usr/lib/perl5/5.8.5/i386-linux-thread-multi
    /usr/lib/perl5/5.8.5
    /usr/lib/perl5/site_perl/5.8.5/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.4/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.3/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.2/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.1/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.5
    /usr/lib/perl5/site_perl/5.8.4
    /usr/lib/perl5/site_perl/5.8.3
    /usr/lib/perl5/site_perl/5.8.2
    /usr/lib/perl5/site_perl/5.8.1
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.5/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.4/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.3/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.2/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.1/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.5
    /usr/lib/perl5/vendor_perl/5.8.4
    /usr/lib/perl5/vendor_perl/5.8.3
    /usr/lib/perl5/vendor_perl/5.8.2
    /usr/lib/perl5/vendor_perl/5.8.1
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.5:
    HOME=/home/simonf
    LANG=en_US.iso885915
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/kerberos/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/games:/usr/java2/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/java2/bin:/usr/java2/bin
    PERL5LIB=/opt/software/perl/lib:/opt/software/perl/lib/perl5/site_perl
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

On Mon\, Jan 24\, 2005 at 03&#8203;:39&#8203;:52PM \-0000\, simonf@&#8203;cshl\.edu \(via RT\) wrote&#8203;:
> 
> The following modules produce this error message&#8203;:
> 
> \-\->perl \-I\. \-MA
> Unknown error
> Compilation failed in require\.
> BEGIN failed\-\-compilation aborted\.
> 
> A\.pm
> \*\*\*
> package A;
> 
> use strict;
> 
> use base C;
> 
> 1;
> 
> \*\*\*
> 
> C\.pm
> \*\*\*
> package C;
> 
> use strict;
> 
> 1;
> 
> \*\*\*
> 
> As far as I can tell\, the name of the second package does not matter\.
> What matters is that you omit quotes around its name in 'use base'\.
>

Looking at the perldoc for use\, anything after the module name must be a list\.
So\, 

use base "C";

is correct\, and 

use base C;

should give you a compile error as you demonstrated above\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tamias

On Mon\, Jan 24\, 2005 at 10&#8203;:37&#8203;:16AM \-0600\, Steve Peters wrote&#8203;:

> Looking at the perldoc for use\, anything after the module name must be a list\.
> So\, 
> 
> use base "C";
> 
> is correct\, and 
> 
> use base C;
> 
> should give you a compile error as you demonstrated above\.

I think the complaint was that the error message is "Unknown error" rather
than something more descriptive\, like "Bareword not allowed after module
name in use statement"\.

Anyway\, C\<C> is a list; it's a list containing a single bareword\.  The
documentation doesn't seem to make it clear that the list cannot contain
barewords\.

Ronald

**Comment by p5pRT** [comment]
### From simonf@cshl.edu

Hi\,

Of course it's incorrect\. My problem is that it should give me a more 
descriptive compile error\.

Simon

Steve Peters via RT wrote&#8203;:
> On Mon\, Jan 24\, 2005 at 03&#8203;:39&#8203;:52PM \-0000\, simonf@&#8203;cshl\.edu \(via RT\) wrote&#8203;:
> 
>>The following modules produce this error message&#8203;:
>>
>>\-\->perl \-I\. \-MA
>>Unknown error
>>Compilation failed in require\.
>>BEGIN failed\-\-compilation aborted\.
>>
>>A\.pm
>>\*\*\*
>>package A;
>>
>>use strict;
>>
>>use base C;
>>
>>1;
>>
>>\*\*\*
>>
>>C\.pm
>>\*\*\*
>>package C;
>>
>>use strict;
>>
>>1;
>>
>>\*\*\*
>>
>>As far as I can tell\, the name of the second package does not matter\.
>>What matters is that you omit quotes around its name in 'use base'\.
>>
> 
> 
> Looking at the perldoc for use\, anything after the module name must be a list\.
> So\, 
> 
> use base "C";
> 
> is correct\, and 
> 
> use base C;
> 
> should give you a compile error as you demonstrated above\.
> 
> 
> 

\-\- 

Simon \(Vsevolod ILyushchenko\)   simonf@&#8203;cshl\.edu
&nbsp;                http&#8203;://www\.simonf\.com

Terrorism is a tactic and so to declare war on terrorism
is equivalent to Roosevelt's declaring war on blitzkrieg\.

Zbigniew Brzezinski\, U\.S\. national security advisor\, 1977\-81

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jan 24\, 2005 at 12&#8203;:43&#8203;:54PM \-0500\, Ronald J Kimball wrote&#8203;:
> On Mon\, Jan 24\, 2005 at 10&#8203;:37&#8203;:16AM \-0600\, Steve Peters wrote&#8203;:
> 
> > Looking at the perldoc for use\, anything after the module name must be a list\.
> > So\, 
> > 
> > use base "C";
> > 
> > is correct\, and 
> > 
> > use base C;
> > 
> > should give you a compile error as you demonstrated above\.
> 
> I think the complaint was that the error message is "Unknown error" rather
> than something more descriptive\, like "Bareword not allowed after module
> name in use statement"\.

This appears to be fixed in bleedperl&#8203;:

&nbsp;    $ \./perl \-I/home/davem/tmp/x \-Ilib \-MA
&nbsp;    Bareword "C" not allowed while "strict subs" in use at /home/davem/tmp/x/A\.pm line 5\.
&nbsp;    Compilation failed in require\.
&nbsp;    BEGIN failed\-\-compilation aborted\.
&nbsp;    $

somewhere between changes 19162 and 19306; probably 19170 judging by the
log entry&#8203;:

\[ 19170\] By&#8203;: rgs                                   on 2003/04/08  19&#8203;:47&#8203;:05
&nbsp;        Log&#8203;: Subject&#8203;: \[PATCH\] Re&#8203;: Error&#8203;: Unknown error
&nbsp;             From&#8203;: Enache Adrian \<enache@&#8203;rdslink\.ro>
&nbsp;             Date&#8203;: Tue\, 8 Apr 2003 07&#8203;:12&#8203;:13 \+0300

Looks like it never got integrated into the 5\.8\.x branch\.


\-\- 
"Do not dabble in paradox\, Edward\, it puts you in danger of fortuitous
it\." \-\- Lady Croom \- Arcadia

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
